jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/Utf8Appendable.java

```diff
@@ -327,40 +331,159 @@ public boolean isUtf8SequenceComplete()
+     * @return the UTF-8 String
+     * @throws NotUtf8Exception if String has replacement characters due to bad UTF-8 sequences
+     * @see #finish()
+     */
+    public String getString(boolean throwOnReplacement)
+    {
+        if (throwOnReplacement && hasReplacements())
         {
-            if (_coderErrorAction == CodingErrorAction.REPLACE)
-            {
-                try
-                {
-                    _appendable.append(REPLACEMENT);
-                }
-                catch (IOException e)
-                {
-                    throw new RuntimeException(e);
-                }
-                _hasReplacements = true;
-            }
-            _codep = 0;
-            _state = UTF8_ACCEPT;
-            if (_coderErrorAction == CodingErrorAction.REPORT)
-                throw new NotUtf8Exception("incomplete UTF8 sequence");
+            throw new NotUtf8Exception("String has invalid UTF-8 sequences");
+        }
+        return _appendable.toString();
+    }
+
+    /**
+     * Get the String as it exists currently.
+     *
+     * <p>
+     *     This will get the String as it exists currently, without including any trailing incomplete UTF-8 sequences.
+     * </p>
+     * <p>
+     *     Use {@link #finish()} to complete any incomplete UTF-8 sequences before calling this method.
+     * </p>
+     *
+     * @return the String as it exists currently, without throwing an Exception
+     * @see #finish()
+     */
+    public String getString()
+    {
+        return getString(false);
+    }
+
+    /**
+     * <p>
+     * Take the String from the {@link Appendable}, taking care to finish any incomplete UTF-8 sequences first.
+     * </p>
+     *
+     * <p>
+     * Calls to this method will {@link #reset()} this {@link Utf8Appendable}.
+     * </p>
+     *
+     * @return the String buffer from the appendable
+     * @throws CharacterCodingException if unable to encode the input bytes (eg: bad UTF-8 sequences)
+     */
+    @Override
+    public String takeString() throws CharacterCodingException
+    {
+        String str = null;
+        try
+        {
+            finish();
+            str = getString(true);
         }
+        catch (NotUtf8Exception e)
+        {
+            throw (CharacterCodingException)new CharacterCodingException().initCause(e);
+        }
+        catch (RuntimeException e)
+        {
+            throw (CharacterCodingException)new CharacterCodingException().initCause(e.getCause());
+        }
+        finally
+        {
+            reset();
+        }
+        return str;
+    }
+
+    /**
+     * Take the complete UTF-8 string and reset the internal buffer.
+     *
+     * @param throwOnReplacement throw {@link NotUtf8Exception} if input had characters replaced.
+     * @return the UTF-8 so far, including any final incomplete UTF-8 sequences.
+     */
+    public String takeFinishedString(boolean throwOnReplacement)
```

Comment:
I can't see the difference between `takeString()` (which calls `finish()`) and `takeFinishedString()`???